### PR TITLE
Fix pointing at arg when cause is outside of call

### LINKED ISSUE #66923: Error message blames wrong line
The code below blames the last line for the error where in fact error is in the previous one. 
The error message makes no sense because `x4` type is specified explicitly and no iterator is present anymore.

```
fn main() {
    let v = vec![1_f64, 2.2_f64];
    let mut fft: Vec<Vec<f64>> = vec![];

    let x1: &[f64] = &v;
    let x2 = x1.into_iter();
    let x3 = x2.rev();
    let x4: Vec<f64> = x3.collect();    // <-- mistake is here
    fft.push(x4);                                 // <-- but error blames this line
}
```

```
--> src/main.rs:9:14
  |
9 |     fft.push(x4);
  |              ^^ a collection of type `std::vec::Vec<f64>` cannot be built from `std::iter::Iterator<Item=&f64>`
```

### PR DESCRIPTION
**Pull request by VirrageS**
Closes: #66923

### CODE REVIEW

File: src/test/ui/issues/issue-66923.rs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+fn main() {
```
Comment (Centril) [inline src/test/ui/issues/issue-66923.rs]: Can you please extend the filename with a short hint at what the test is about as well as a comment in the file itself explaining what this is supposed to test?

r=me rollup with that done.

File: src/test/ui/issues/issue-66923-show-error-for-correct-call.stderr [hunk 2]
```diff
@@ -0,0 +1,19 @@
+error[E0277]: a collection of type `std::vec::Vec<f64>` cannot be built from an iterator over elements of type `&f64`
+  --> $DIR/issue-66923-show-error-for-correct-call.rs:6:39
```
Comment (lzutao) [inline src/test/ui/issues/issue-66923-show-error-for-correct-call.stderr]: Bless stderr to match test file line.

Comment (VirrageS) [reply src/test/ui/issues/issue-66923-show-error-for-correct-call.stderr]: Thanks!

**Comment by JohnTitor** [comment]
Seems failed in https://github.com/rust-lang/rust/pull/67042#issuecomment-561941965
@bors r-

**Comment by lzutao** [comment]
@JohnTitor I think it shall be #66764.

**Comment by JohnTitor** [comment]
@lzutao Indeed! So rebase will be needed once it is merged.

**Comment by lzutao** [comment]
@VirrageS You might want to rebase agains latest master and bless the test.

**Comment by VirrageS** [comment]
@lzutao done :)

**Comment by VirrageS** [comment]
`src/test/ui/const-generics/array-impls/core-traits-no-impls-length-33.rs` seems to fail and I don't think I know how to fix it. Mostly because from similar code we expect different behaviour:

```
fn foo() {
    let mut set = HashSet::new(); // don't point here!
    set.insert([0_usize; 33]); // <- point here
}
```
here we have 2 errors. One is connected to argument itself `[0_usize; 33]` and the second one (I think) is connected to coerced type of hash set which tries to be `HashSet<[0_usize; 33]>` and because it cannot be created (because of `std::array::LengthAtMost32` trait constraints)  we have an error. With current implementation we point both errors to argument (which is good) so they collapse into single one that is reported but with my implementation we have two different errors where one of them (the one which points to 2nd line) can be very confusing.

```
fn foo() {
    let v = vec![1_f64, 2.2_f64];
    let mut fft: Vec<Vec<f64>> = vec![];

    let x1: &[f64] = &v;
    let x2 = x1.into_iter().collect::<Vec<f64>>(); // <- point here
    fft.push(x2); // don't point here!
}
```

But here we also have similar situation but now we don't want to point to argument but rather to type mismatch which happens where we want to create `Vec<f64>` out of `[&64]`.

**Comment by VirrageS** [comment]
Since I don't have enough knowledge to even try to fix these discrepancies I'm closing this MR.

**Comment by VirrageS** [comment]
The only question is why it worked/compiled before since `src/test/ui/const-generics/array-impls/core-traits-no-impls-length-33.rs` is pretty old stuff.

**Comment by lzutao** [comment]
Wouldn't you also bless that test `src/test/ui/const-generics/array-impls/core-traits-no-impls-length-33.rs` is better?

**Comment by VirrageS** [comment]
The changes are not showing up after I closed the MR but you can check stderr for the blessed test here: https://github.com/VirrageS/rust/commit/e305bf8bc8cecaf93080f07962a4f763cd66f5ce#diff-83c38304ce39064b906661170076d85e 

```
		 error[E0277]: arrays only have std trait implementations for lengths 0..=32
   --> $DIR/core-traits-no-impls-length-33.rs:8:19
    |
 LL |     let mut set = HashSet::new();
    |                   ^^^^^^^^^^^^ the trait `std::array::LengthAtMost32` is not implemented for `[usize; 33]`
    |
    = note: required because of the requirements on the impl of `std::cmp::Eq` for `[usize; 33]`
    = note: required by `std::collections::HashSet::<T>::new`
```

This is rather confusing than helpful :C

**Comment by estebank** [comment]
I do think that the new output in that file is not optimal, but good enough for the change to land, because it isn't misleading, just verbose.

**Comment by VirrageS** [comment]
I can definitelly reopen the new MR (I cannot reopen this for some reason), maybe we can start a new discussion there. I will do that later today.

**Comment by estebank** [comment]
> (I cannot reopen this for some reason)

Github doesn't let you reopen a pull request that has been closed and _then_ has had it's source branch updated.

**Comment by VirrageS** [comment]
I see :( I've reopened it here: #67354

**Comment by eddyb** [comment]
> Github doesn't let you reopen a pull request that has been closed and then has had it's source branch updated.

For future reference, you *can* fix this by force-pushing the previous commit (I believe the syntax is `git push -f origin e2e4801:blame-wrong-line`), reopening the PR, then force-pushing your changes.

**Pull request closed without merging**